hpke/kembase.go

```diff
@@ -0,0 +1,234 @@
+package hpke
+
+import (
+	"crypto"
+	_ "crypto/sha256" // linking sha256 packages.
+	_ "crypto/sha512" // linking sha512 packages.
+	"encoding"
+	"encoding/binary"
+	"io"
+
+	"github.com/cloudflare/circl/kem"
+	"golang.org/x/crypto/hkdf"
+)
+
+type dhKem interface {
+	sizeDH() int
+	calcDH(dh []byte, sk kem.PrivateKey, pk kem.PublicKey) error
+	GenerateKey() (kem.PublicKey, kem.PrivateKey, error)
+	DeriveKey(seed []byte) (kem.PublicKey, kem.PrivateKey)
+	UnmarshalBinaryPrivateKey(data []byte) (kem.PrivateKey, error)
+	UnmarshalBinaryPublicKey(data []byte) (kem.PublicKey, error)
+}
+
+type kemBase struct {
+	id   KemID
+	name string
+	crypto.Hash
+	dh dhKem
+}
+
+func (k kemBase) Name() string       { return k.name }
+func (k kemBase) SharedKeySize() int { return k.Hash.Size() }
+
+func (k kemBase) getSuiteID() (sid [5]byte) {
+	sid[0], sid[1], sid[2] = 'K', 'E', 'M'
+	binary.BigEndian.PutUint16(sid[3:5], uint16(k.id))
+	return
+}
+
+func (k kemBase) extractExpand(dh, kemCtx []byte) []byte {
+	eaePkr := k.labeledExtract(nil, []byte("eae_prk"), dh)
+	return k.labeledExpand(
+		eaePkr,
+		[]byte("shared_secret"),
+		kemCtx,
+		uint16(k.Size()),
+	)
+}
+
+func (k kemBase) labeledExtract(salt, label, info []byte) []byte {
+	suiteID := k.getSuiteID()
+	labeledIKM := append(append(append(append(
+		make([]byte, 0, len(versionLabel)+len(suiteID)+len(label)+len(info)),
+		versionLabel...),
+		suiteID[:]...),
+		label...),
+		info...)
+	return hkdf.Extract(k.New, labeledIKM, salt)
+}
+
+func (k kemBase) labeledExpand(prk, label, info []byte, l uint16) []byte {
+	suiteID := k.getSuiteID()
+	labeledInfo := make(
+		[]byte,
+		2,
+		2+len(versionLabel)+len(suiteID)+len(label)+len(info),
+	)
+	binary.BigEndian.PutUint16(labeledInfo[0:2], l)
+	labeledInfo = append(append(append(append(labeledInfo,
+		versionLabel...),
+		suiteID[:]...),
+		label...),
+		info...)
+	b := make([]byte, l)
+	rd := hkdf.Expand(k.New, prk, labeledInfo)
+	if n, err := io.ReadFull(rd, b); err != nil || n != int(l) {
+		panic(err)
+	}
+	return b
+}
+
+func (k kemBase) AuthEncapsulate(
+	pkr kem.PublicKey,
+	sks kem.PrivateKey,
+) (ct []byte, ss []byte, err error) {
+	pke, ske, err := k.dh.GenerateKey()
+	if err != nil {
+		return nil, nil, err
+	}
+	return k.authEncap(pkr, sks, pke, ske)
+}
+
+func (k kemBase) Encapsulate(
+	pkr kem.PublicKey,
+) (ct []byte, ss []byte, err error) {
+	pke, ske, err := k.dh.GenerateKey()
```

Comment:
The test vectors validate `EncapsulateDeterminstically()`, but not `Encapsulate()`. To increase test coverage, I would do the following. Instead of calling `GenerateKey()` here, I would generate a seed and pass it to `DeriveKey()`, similar to `EncapsulateDeterministically` below.

Reply:
updated